internal/component/kafka/consumer.go

```diff
@@ -53,10 +59,113 @@ func (consumer *consumer) ConsumeClaim(session sarama.ConsumerGroupSession, clai
+	b backoff.BackOff) error {
+	handlerConfig, err := consumer.k.GetTopicBulkHandlerConfig(claim.Topic())
+	if err != nil {
+		return fmt.Errorf("error getting bulk handler config for topic %s: %w", claim.Topic(), err)
+	}
+	ticker := time.NewTicker(time.Duration(handlerConfig.SubscribeConfig.MaxBulkAwaitDurationMilliSeconds) * time.Millisecond)
+	defer ticker.Stop()
+	messages := make([]*sarama.ConsumerMessage, 0, handlerConfig.SubscribeConfig.MaxBulkCount)
+	for {
+		select {
+		case <-session.Context().Done():
+			return consumer.flushBulkMessages(claim, messages, session, handlerConfig.Handler, b)
+		case message := <-(claim.Messages()):
+			if message != nil {
+				messages = append(messages, message)
+				if len(messages) >= handlerConfig.SubscribeConfig.MaxBulkCount {
+					return consumer.flushBulkMessages(claim, messages, session, handlerConfig.Handler, b)
+				}
+			}
+		case <-ticker.C:
+			return consumer.flushBulkMessages(claim, messages, session, handlerConfig.Handler, b)
+		}
+	}
+}
 
+func (consumer *consumer) flushBulkMessages(claim sarama.ConsumerGroupClaim,
+	messages []*sarama.ConsumerMessage, session sarama.ConsumerGroupSession,
+	handler BulkEventHandler, b backoff.BackOff) error {
+	if len(messages) > 0 {
+		if consumer.k.consumeRetryEnabled {
+			if err := retry.NotifyRecover(func() error {
+				return consumer.doBulkCallback(session, messages, handler, claim.Topic())
+			}, b, func(err error, d time.Duration) {
+				consumer.k.logger.Warnf("Error processing Kafka bulk messages: %s. Error: %v. Retrying...", claim.Topic(), err)
+			}, func() {
+				consumer.k.logger.Infof("Successfully processed Kafka message after it previously failed: %s", claim.Topic())
+			}); err != nil {
+				consumer.k.logger.Errorf("Too many failed attempts at processing Kafka message: %s. Error: %v.", claim.Topic(), err)
+			}
+		} else {
+			err := consumer.doBulkCallback(session, messages, handler, claim.Topic())
+			if err != nil {
+				consumer.k.logger.Errorf("Error processing Kafka message: %s. Error: %v.", claim.Topic(), err)
+			}
+			return err
+		}
+	}
 	return nil
 }
 
+func (consumer *consumer) doBulkCallback(session sarama.ConsumerGroupSession,
+	messages []*sarama.ConsumerMessage, handler BulkEventHandler, topic string) error {
+	consumer.k.logger.Debugf("Processing Kafka bulk message: %s", topic)
+	messageValues := make([]KafkaBulkMessageEntry, (len(messages)))
+
+	for i, message := range messages {
+		if message != nil {
+			metadata := make(map[string]string, len(message.Headers))
+			if message.Headers != nil {
+				for _, t := range message.Headers {
+					metadata[string(t.Key)] = string(t.Value)
+				}
+			}
+			var entryIDStr string
+			entryID, entryIDErr := uuid.NewRandom()
+			if entryIDErr != nil {
+				consumer.k.logger.Warnf("Failed to generate UUID for sending as entryID in message for bulk subscribe event on topic: %s. Error: %v.", topic, entryIDErr)
+				entryIDStr = strconv.Itoa(i)
```

Comment:
Why not simply default to this? What is the value to try to generate an UUID for a temporary id?

Reply:
In a particular bulk subscribe call, entryID needs to be unique - hence, I can't make it default. Ideally, NewRandom() will not error out, it is just an extra bit of precaution taken.

Reply:
As discussed offline, I got your intent - Why not use index itself an entryID ad uuid seems overkill? I am perfectly fine with that, as I myself also had earlier assigned index as entryID. I have changed to that.